### PR TITLE
Switch to PublicApiAnalyzers

### PR DESCRIPTION
**Pull request by rainersigwald**
Abandon GenAPI for the new hotness in public API change tracking, which

1. Is a Roslyn analyzer.
2. Works cross-platform.
3. Has clear messages.
4. Actually tracks the API, instead of making us do it manually on review.

Based off the `vs17.0` branch so I didn't accidentally canonize new API.

### CODE REVIEW

**Comment by rainersigwald** [comment]
@ladipro, I'd especially like your review on this because of the StringTools API-tracking stuff.

**Review by jeffkl — APPROVED** [review]
GenAPI has served us well but I'm excited to see us adopt more evolved solution!

**Comment by rainersigwald** [comment]
The merge I just pushed will demonstrate what the errors look like in CI (that's totally why I'm doing it, nothing to do with laziness).

**Comment by rainersigwald** [comment]
Not too bad IMO:

image

Now I'll push a better merge.

**Comment by rainersigwald** [comment]
> Now I'll push a better merge.

Let my failure to do this be a WARNING and a CAVEAT to this: I used "fix all in project" and it only did the first TF of that project. You can instead use "fix all in **solution**" to get everything (per https://github.com/dotnet/roslyn-analyzers/issues/4954#issuecomment-669135752).

File: src/Directory.Build.targets [hunk 1]
```diff
@@ -74,23 +74,21 @@
...
-  </PropertyGroup>
-
-  <PropertyGroup Condition="'$(GenerateReferenceAssemblySource)' == 'true' and $([MSBuild]::IsOSPlatform('windows'))">
+  <PropertyGroup Condition="'$(GenerateReferenceAssemblySource)' == 'true'">
     <GenAPIAssemblyName>$(AssemblyName)</GenAPIAssemblyName>
     <GenAPIAssemblyName Condition="'$(GenAPIAssemblyName)' == ''">$(MSBuildProjectName)</GenAPIAssemblyName>
     <GenAPIShortFrameworkIdentifier Condition="$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETFramework'">net</GenAPIShortFrameworkIdentifier>
     <GenAPIShortFrameworkIdentifier Condition="$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETStandard'">netstandard</GenAPIShortFrameworkIdentifier>
     <GenAPIShortFrameworkIdentifier Condition="$([MSBuild]::GetTargetFrameworkIdentifier('$(TargetFramework)')) == '.NETCoreApp'">netstandard</GenAPIShortFrameworkIdentifier>
-    <GenAPITargetPath>$(RepoRoot)ref\$(GenAPIAssemblyName)\$(GenAPIShortFrameworkIdentifier)\$(GenAPIAssemblyName).cs</GenAPITargetPath>
```
Comment (BenVillalobos) [inline src/Directory.Build.targets:89]: I noticed GenAPITargetPath wasn't used before. Was it getting picked up somewhere else in the build, or was this an unused property?

**Review by BenVillalobos — APPROVED** [review]

File: src/Directory.Build.targets (hunk 1 above)
Comment (rainersigwald) [reply src/Directory.Build.targets:89]: It was used in the GenAPI package's targets somewhere.

Comment (BenVillalobos) [reply src/Directory.Build.targets:89]: Got it, and we're not using them anymore so should we get rid of some of these variables? Or proactively rename them to match the analyzer we're moving to.

Consider this a nit though.

**Review by ladipro — APPROVED** [review]
Looks great, thank you! What is supposed to go into the `Unshipped` files? The [documentation](https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) just says they must exist; no details given.

**Comment by rainersigwald** [comment]
> What is supposed to go into the `Unshipped` files?

Lol, yeah, we should write up some docs. The PublicApiAnalyzers process goes like this:

1. The analyzer keeps the `PublicAPI.*.txt` files updated. 
2. New API surface goes into `PublicAPI.Unshipped.txt`.
3. At release time, we must manually promote the `Unshipped` public API to `Shipped`

The idea is to distinguish between new public API surface in a given release (can still be changed if we figure out a better API shape, because we don't have to maintain compat with daily builds) and shipped public API surface (which we shouldn't change because someone might be using it).

I'll write up a doc.

**Pull request merged**